File: hw3/predictive_rec_descent.py

from typing import List, Tuple

precedences = {'or': 1, 'and': 2, 'not': 3, 'id': 4}

def is_id(tok: str) -> bool:
    return isinstance(tok, str) and len(tok) == 1 and 'a' <= tok <= 'z'

# Helper Functions

# paren_L(E = child_tokens, E.prec = child_prec, p = p)
def paren_L(child_tokens: List[str], child_prec: int, p: int) -> List[str]:
    if child_prec < p:
        return ['('] + child_tokens + [')']
    return child_tokens

# paren_R(E = child_tokens, E.prec = child_prec, p = p)
def paren_R(child_tokens: List[str], child_prec: int, p: int) -> List[str]:
    if child_prec <= p:
        return ['('] + child_tokens + [')']
    return child_tokens

# paren_unary(E = child_tokens, E.prec = child_prec, p = p)
def paren_unary(child_tokens: List[str], child_prec: int, p: int) -> List[str]:
    if child_prec < p:
        return ['('] + child_tokens + [')']
    return child_tokens

# Parser Class 
class Parser:
    def __init__(self, tokens: List[str]):
        self.toks = tokens
        self.i = 0

    # Lookahead function: looks at the next token without consuming it
    def lookahead(self):
        if self.i >= len(self.toks):
            return None
        else:
            return self.toks[self.i]

    # Match function: consumes the expected token if it matches the lookahead
    def match(self, expected: str):
        if self.lookahead() == expected:
            self.i += 1
        else:
            raise ValueError(f"PARSE_ERROR: expected {expected}, saw {self.lookahead()}")

    # Check if at the end of the token list
    def at_end(self) -> bool:
        return self.i >= len(self.toks)

    # Start function: entry point for parsing
    def start(self) -> List[str]:
        result_tokens, _prec = self.expr()
        if self.lookahead() == '$$':
            self.match('$$')
        if not self.at_end():
            raise ValueError("PARSE_ERROR: extra tokens at end")
        return result_tokens

    # Expr function: handles the grammar rules
    def expr(self) -> Tuple[List[str], int]:
        t = self.lookahead() # lookahead token
        # Based on the lookahead token, decide which production to use
        if t == 'or': 
            self.match('or')
            L_toks, L_p = self.expr() # parse left expr
            R_toks, R_p = self.expr() # parse right expr
            p = precedences['or'] 
            L = paren_L(L_toks, L_p, p) # add parentheses to left if needed
            R = paren_R(R_toks, R_p, p) # add parentheses to right if needed
            return (L + ['or'] + R, p)

        elif t == 'and': 
            self.match('and')
            L_toks, L_p = self.expr() # parse left expr
            R_toks, R_p = self.expr() # parse right expr
            p = precedences['and']
            L = paren_L(L_toks, L_p, p) # add parentheses to left if needed
            R = paren_R(R_toks, R_p, p) # add parentheses to right if needed
            return (L + ['and'] + R, p)

        elif t == 'not':
            self.match('not') 
            S_toks, S_p = self.expr() # parse sub expr
            p = precedences['not']
            S = paren_unary(S_toks, S_p, p) # add parentheses to sub expr if needed
            return (['not'] + S, p)

        elif is_id(t):
            self.match(t)  # consume the id
            return ([t], precedences['id'])

        else:
            raise ValueError("PARSE_ERROR: expected 'or' | 'and' | 'not' | id")
    
def parse(tokens: List[str]) -> str:
    toks = tokens
    try:
        p = Parser(toks)
        result_tokens = p.start()
        print(' '.join(result_tokens))
    except Exception:
        print('')

File: hw3/predictive_rec_descent.pl
% entry point
parse(Tokens, Result) :-
    expr(Tokens, Rest, ResTokens, _Prec),
    % accept either exact consumption or a single '$$' end marker
    ( Rest = []
    ; Rest = ['$$'] ),
    atomic_list_concat(ResTokens, ' ', Result).

% productions
% expr -> or  expr expr
% expr -> and expr expr
% expr -> not expr
% expr -> id

% Based on current(lookahead) token, decide which production to use

% or case
expr([or|T0], Rest, Res, 1) :-
    expr(T0, T1, L_toks, L_p), % parse left expr
    expr(T1, Rest0, R_toks, R_p), % parse right expr
    paren_L(L_toks, L_p, 1, L1), % add parens to left if needed
    paren_R(R_toks, R_p, 1, R1), % add parens to right if needed
    append(L1, [or|R1], Res),
    Rest = Rest0.

% and case
expr([and|T0], Rest, Res, 2) :-
    expr(T0, T1, L_toks, L_p), % parse left expr
    expr(T1, Rest0, R_toks, R_p), % parse right expr
    paren_L(L_toks, L_p, 2, L1), % add parens to left if needed
    paren_R(R_toks, R_p, 2, R1), % add parens to right if needed
    append(L1, [and|R1], Res),
    Rest = Rest0.

% not case
expr([not|T0], Rest, Res, 3) :-
    expr(T0, Rest, S_toks, S_p), % parse sub expr
    paren_un(S_toks, S_p, 3, S1), % add parens to sub expr if needed
    Res = [not|S1].

% base case: id
expr([Id|Rest], Rest, [Id], 4) :- 
    is_id(Id).

% Wrap atom in parantheses if child precedence 
% is less than parent precedence

paren_L(Toks, ChildPrec, ParentPrec, Res) :-
    ChildPrec < ParentPrec, !,
    wrap_parens(Toks, Res).
paren_L(Toks, _ChildPrec, _ParentPrec, Toks).

paren_R(Toks, ChildPrec, ParentPrec, Res) :-
    ChildPrec =< ParentPrec, !,
    wrap_parens(Toks, Res).
paren_R(Toks, _ChildPrec, _ParentPrec, Toks).

paren_un(Toks, ChildPrec, ParentPrec, Res) :-
    ChildPrec < ParentPrec, !,
    wrap_parens(Toks, Res).
paren_un(Toks, _ChildPrec, _ParentPrec, Toks).

wrap_parens(T, ['('|T2]) :-
    append(T, [')'], T2).

% checks if an atom is a valid id (a-z)
is_id(Id) :-
    atom(Id),
    atom_length(Id, 1),
    atom_chars(Id, [C]),
    member(C, [a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z]).